internal/controller/mariadb_controller.go

```diff
@@ -341,6 +348,53 @@ func (r *MariaDBReconciler) reconcileStatefulSet(ctx context.Context, mariadb *m
 	return ctrl.Result{}, nil
 }
 
+func (r *MariaDBReconciler) reconcilePodLabels(ctx context.Context, mariadb *mariadbv1alpha1.MariaDB) (ctrl.Result, error) {
+	podList := corev1.PodList{}
+	listOpts := &client.ListOptions{
+		LabelSelector: klabels.SelectorFromSet(
+			labels.NewLabelsBuilder().
+				WithMariaDBSelectorLabels(mariadb).
+				Build(),
+		),
+		Namespace: mariadb.GetNamespace(),
+	}
+	if err := r.List(ctx, &podList, listOpts); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error listing Pods: %v", err)
+	}
+
+	for _, pod := range podList.Items {
+		var role = "replica"
+
+		if pod.Status.PodIP == "" || pod.Spec.NodeName == "" {
+			continue
+		}
+		podIndex, err := sts.PodIndex(pod.Name)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error getting Pod '%s' index: %v", pod.Name, err)
+		}
+
+		if mariadb.Status.CurrentPrimaryPodIndex != nil && *podIndex == *mariadb.Status.CurrentPrimaryPodIndex {
```

Comment:
Instead of doing this check here could we do this at the begining of `reconcilePodLabels` the function:

```go
if mariadb.Status.CurrentPrimaryPodIndex == nil {
    return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
}
``` 

`CurrentPrimaryPodIndex` will be set eventually, but without this early return we will be marking all pods as `replica`, which is not very precise. Then we can update the current statement to just:

```suggestion
		if  *podIndex == *mariadb.Status.CurrentPrimaryPodIndex {
```

Doing as suggested should do the trick.

Reply:
Thanks, I haven't seen that happen during testing, primary was always coming first before I saw replica labels being applied but doesn't hurt to be more precise :)